File: vs2013_test/test_table/hulib.cpp
#include <string.h>
#include "hulib.h"
#include "table_mgr.h"

bool HuLib::get_hu_info(char* const hand_cards, Wave* const waves, char self_card_index, char other_card_index, int gui_index)
{
    char hand_cards_tmp[34];
    memcpy(hand_cards_tmp, hand_cards, 34);

    if(self_card_index != 34)
    {
        hand_cards_tmp[self_card_index]++;
    }
    else if(other_card_index != 34)
    {
        hand_cards_tmp[other_card_index]++;
    }

    int gui_num = hand_cards_tmp[gui_index];
    hand_cards_tmp[gui_index] = 0;

	return split(hand_cards_tmp, gui_num);
}

bool check(int gui, int eye_num, int gui_num, int& gui_sum) {
	if (gui < 0) return false;

	gui_sum += gui;
	if (gui_sum > gui_num) return false;

	if (eye_num == 0) return true;

	return gui_sum + (eye_num - 1) <= gui_num;
}

bool HuLib::split(char* const cards, int gui_num)
{
	int eye_num = 0;
	int gui_sum = 0;
	int gui = 0;
	
	gui = _split(cards, gui_num, 0, 8, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;
	
	gui = _split(cards, gui_num-gui_sum, 9, 17, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;

	gui = _split(cards, gui_num-gui_sum, 18, 26, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;

	gui = _split(cards, gui_num-gui_sum, 27, 33, false, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;

	if (eye_num == 0)
	{
		return gui_sum + 2 <= gui_num;
	}

	return true;
}

int HuLib::_split(char* const cards, int gui_num, int min, int max, bool chi, int& eye_num)
{
    int key = 0;
    int num = 0;

    for(int i = min; i <= max; ++i)
    {
        key = key*10 + cards[i];
        num = num + cards[i];
    }

	if (num == 0) return 0;

	for (int i = 0; i <= gui_num; ++i)
	{
		int yu = (num + i) % 3;
		if (yu == 1) continue;
		bool eye = (yu == 2);
		if (TableMgr::get_instance()->check(key, i, eye, chi))
		{
			if (eye) eye_num++;
			return i;
		}
	}

    return -1;
}

File: mjlib_lua/gui_table/hulib.lua
local table_mgr = require "table_mgr"

local M = {}

function M.get_hu_info(cards, gui_index)
    local hand_cards = {}
    for i,v in ipairs(cards) do
        hand_cards[i] = v
    end

    local gui_num = 0
    if gui_index > 0 then
        gui_num = hand_cards[gui_index]
        hand_cards[gui_index] = 0
    end

    return M.check(hand_cards, gui_num)
end

function M.check(cards, gui_num)
	local total_need_gui = 0
	local eye_num = 0
	for i=0,3 do
		local from = i*9 + 1
		local to = from + 8
		if i == 3 then
			to = from + 6
		end
		
		local need_gui, eye = M.get_need_gui(cards, from, to, i<3, gui_num)
		if not need_gui then
		    return false
		end
		total_need_gui = total_need_gui + need_gui
		if eye then
			eye_num = eye_num + 1
		end
	end
	
	if eye_num == 0 then
		return total_need_gui + 2 <= gui_num
	elseif eye_num == 1 then
		return total_need_gui <= gui_num
	elseif
		return total_need_gui + eye_num - 1 <= gui_num
	end
end

function M.get_need_gui(cards, from, to, chi, gui_num)
	local num = 0
	local key = 0
	for i=from,to do
		key = key * 10 + cards[i]
		num = num + cards[i]
	end
	
	if num == 0 then
	    return 0, false
	end

    for i=0, gui_num do
        local yu = (num + i)%3
        if yu ~= 1 then
            local eye = (yu == 2)
            if table_mgr:check(key, i, eye, chi) then
                return i, eye
            end
        end
    end
end

return M
